addons/account/partner.py

```diff
@@ -241,11 +241,28 @@ def _debit_search(self, cr, uid, obj, name, args, context=None):
     def _invoice_total(self, cr, uid, ids, field_name, arg, context=None):
         result = {}
         account_invoice_report = self.pool.get('account.invoice.report')
-        for partner in self.browse(cr, uid, ids, context=context):
-            domain = [('partner_id', 'child_of', partner.id), ('state', 'not in', ['draft', 'cancel'])]
+        user = self.pool['res.users'].browse(cr, uid, uid, context=context)
+        for partner_id in ids:
+            # first read invoices in the user's currency
+            all_partner_ids = self.pool['res.partner'].search(
+                cr, uid, [('id', 'child_of', partner_id)], context=context)
+            cr.execute(
+                'select sum(price_total) '
+                'from account_invoice_report '
+                'where '
+                'partner_id in %s and state not in %s and currency_id=%s',
+                (tuple(all_partner_ids), tuple(['draft', 'cancel']),
+                 user.company_id.currency_id.id))
+            result[partner_id] = sum(s or 0.0 for s, in cr.fetchall()) or 0.0
```

Comment:
I am concerned this query does not handle multi company correctly, a
scenario which could return more than one record, but shouldn't.

In addons/account/partner.py
<https://github.com/odoo/odoo/pull/6654#discussion_r30240331>:

> @@ -241,11 +241,28 @@ def _debit_search(self, cr, uid, obj, name, args, context=None):
>      def _invoice_total(self, cr, uid, ids, field_name, arg, context=None):
>          result = {}
>          account_invoice_report = self.pool.get('account.invoice.report')
> -        for partner in self.browse(cr, uid, ids, context=context):
> -            domain = [('partner_id', 'child_of', partner.id), ('state', 'not in', ['draft', 'cancel'])]
> +        user = self.pool['res.users'].browse(cr, uid, uid, context=context)
> +        for partner_id in ids:
> +            # first read invoices in the user's currency
> +            all_partner_ids = self.pool['res.partner'].search(
> +                cr, uid, [('id', 'child_of', partner_id)], context=context)
> +            cr.execute(
> +                'select sum(price_total) '
> +                'from account_invoice_report '
> +                'where '
> +                'partner_id in %s and state not in %s and currency_id=%s',
> +                (tuple(all_partner_ids), tuple(['draft', 'cancel']),
> +                 user.company_id.currency_id.id))
> +            result[partner_id] = sum(s or 0.0 for s, in cr.fetchall()) or 0.0

you will get at most only one record so no need for a sum, you can use
fetchone

—
Reply to this email directly or view it on GitHub
<https://github.com/odoo/odoo/pull/6654/files#r30240331>.